pkg/controller/queuejob/queuejob_controller_ex.go

```diff
@@ -934,31 +986,39 @@ func (cc *XController) manageQueueJob(qj *arbv1.AppWrapper) error {
 					}
 				}
 			}
-			if glog.V(10) {
-				if (is_first_deployment) {
-					current_time := time.Now()
-					glog.V(10).Infof("[TTime] %s, %s: WorkerBeforeCreatingResouces", qj.Name, time.Now().Sub(qj.CreationTimestamp.Time))
-					glog.V(10).Infof("[Agent Controller] XQJ %s has Overhead Before Creating Resouces: %s", qj.Name, current_time.Sub(qj.CreationTimestamp.Time))
-				}
-			}
+//			if glog.V(10) {
+//				if (is_first_deployment) {
+//					current_time := time.Now()
+//					glog.V(10).Infof("[TTime] %s, %s: WorkerBeforeCreatingResouces", qj.Name, time.Now().Sub(qj.CreationTimestamp.Time))
+//					glog.V(10).Infof("[Agent Controller] XQJ %s has Overhead Before Creating Resouces: %s", qj.Name, current_time.Sub(qj.CreationTimestamp.Time))
+//				}
+//			}
+			glog.V(4).Infof("[worker-manageQJ] %s 3Delay=%s BeforeDispatchingToEtcd version=%s Status=%+v",
+				qj.Name, metav1.Now().Sub(qj.Status.ControllerFirstTimestamp.Time), qj.ResourceVersion, qj.Status)
+			success := false
 			for _, ar := range qj.Spec.AggrResources.Items {
 				err00 := cc.qjobResControls[ar.Type].SyncQueueJob(qj, &ar)
 				if err00 != nil {
 					glog.V(4).Infof("Error from sync job: %v", err00)
+				} else {
+					success = true
 				}
 			}
-			if glog.V(10) {
-				if (is_first_deployment) {
-					current_time := time.Now()
-					glog.V(10).Infof("[TTime] %s, %s: WorkerAfterCreatingResouces", qj.Name, time.Now().Sub(qj.CreationTimestamp.Time))
-					glog.V(10).Infof("[Agent Controller] XQJ %s has Overhead After Creating Resouces: %s", qj.Name, current_time.Sub(qj.CreationTimestamp.Time))
-				}
+			if success { // set QueueJobStateRunning if at least one resource is successfully dispatched
+				qj.Status.QueueJobState = arbv1.QueueJobStateDispatched
+				glog.V(4).Infof("[worker-manageQJ] %s 4Delay=%s AtLeastOneResourceDispatchedToEtcd version=%s Status=%+v",
+					qj.Name, metav1.Now().Sub(qj.Status.ControllerFirstTimestamp.Time), qj.ResourceVersion, qj.Status)
```

Comment:
@zhangliresearch An appwrapper with no `items` could exists and could not move to the state of `dispatched`.  I don't think there needs to be a check to assign a `dispatched` state.   I think at this point in the logic we would consider the AW dispatch whether no `items` or none of the `items` successfully `synced`.  What are your thoughts?

Reply:
Three points:

1) What do we do with an AppWrapper that has no Items in AggrResources?  Shouldn't we kick them out in the very beginning?  Is there any reason for someone to submit an AppWrapper job with no Items?  Is there any reason if someone does submit a no-Item AppWrapper job, we should keep the job around?  What do we want to do with it?  At minimal, we can check for it up front, mark it complete or delete it.  We should not even let such a job reach this point of the code.  What do you think?

2) err00 := cc.qjobResControls[ar.Type].SyncQueueJob(qj, &ar) submits each Item to etcd.  err00=nil means the submission is successful.  Here, "Dispatched" means qj has been successfully submitted to etcd.  Kubernetes will then perform its actions to get the corresponding pods in a Running state.  We will find out about the Running pods after the cc.qjobResControls[arbv1.ResourceTypePod].UpdateQueueJobStatus(qj) call.  

3) What should we do if the submission of some of the Items failed.  Current code already sets qj.Status.State = arbv1.AppWrapperStateActive without putting it back to the eventQueue.  Should we put qj back to the eventQueue for retry?

Reply:
While this is an corner case Right now the implementation technically supports and AW with no items.  To clarify my question, what behavior is expected to transition into 'Dispatched' state in the following conditions: 
1) AW with no items
2) AW with only non-compute resource demand items (e.g. an AW with a 'Service' resource defined and/or a 'NameSpace resource defined).  We currently support is model.

I think the logic in the PR defines the state transition to 'Dispatched' as all items in the AW have attempted to be 'synced' and at least one item has been successfully 'synced'.  If my understanding of that state definition is correct I'm wondering why differentiate between at least one item and no items?  Do we expect the controller to behave differently in those two conditions?

Reply:
Per our discussion, added handling of incomplete object creation in etcd and set State as Failed. Improved logging information.

Reply:
Awesome.  For now let's go with this implementation as we consider additional behavior to support for partial updates to etcd.  Thank you!